assignments/introductory-assignments.md

```diff
@@ -488,14 +582,767 @@ A digital circuit computer with 2GB of memory can distinguish 2^{17 179 869 184}
+
+The key idea is that both sequences are supposed to have the same semantics. However, `selfie` is executable by a computer, whereas `selfie.c` is not, at least not purposefully, yet `selfie.c` is human-readable and writable in particular. The process of changing `selfie.c` into `selfie` is called *compilation* which is done by a compiler such as the above cc compiler.
+
+[Machine Code](https://en.wikipedia.org/wiki/Machine_code) A sequence of instructions executed directly by a computer's central processing unit (CPU).
+
+[Compiler](https://en.wikipedia.org/wiki/Compiler) A computer program that transforms source code written in a programming language (the source language) into another computer language (the target language), with the latter often having a binary form known as an object or machine code. The most common reason for converting source code is to create an executable program.
+
+Now we have a version of selfie that we can run on our machine.
+
+```sh
+./selfie
+```
+
+**Output**:
+
+```sh
+synopsis: ./selfie { -c { source } | -o binary | [ -s | -S ] assembly | -l binary } [ ( -m | -d | -r | -y ) 0-4096 ... ]
+
+```
+
+Selfie requires using at least one option to do anything useful and therefore responds with its usage pattern and then terminates without doing anything else. 
+
+
+To do something useful, we compile selfie with its own compiler named *starc* that is invoked by the option `-c`.
+
+```sh
+./selfie -c selfie.c
+```
+
+**Output**:
+
+```sh
+./selfie: selfie compiling selfie.c with starc
+./selfie: 293026 characters read in 10278 lines and 1384 comments
+./selfie: with 175311(59.83%) characters in 42447 actual symbols
+./selfie: 401 global variables, 526 procedures, 423 string literals
+./selfie: 2555 calls, 1170 assignments, 86 while, 814 if, 465 return
+./selfie: symbol table search time was 2 iterations on average and 48626 in total
+./selfie: 167872 bytes generated with 38720 instructions and 12992 bytes of data
+./selfie: init:    lui: 2141(5.52%), addi: 14078(36.35%)
+./selfie: memory:  ld: 6474(16.72%), sd: 6080(15.70%)
+./selfie: compute: add: 2944(7.60%), sub: 707(1.82%), mul: 414(1.06%)
+./selfie: compute: divu: 73(0.18%), remu: 29(0.07%)
+./selfie: compare: sltu: 675(1.74%)
+./selfie: control: beq: 904(2.33%), jal: 3667(9.47%), jalr: 526(1.35%)
+./selfie: system:  ecall: 8(0.02%)
+```
+
+
+After compiling `selfie.c` *starc* only stores the machine code internally but does not write it to a file. To do that, we need to use the `-o` option:
+
+```sh
+./selfie -c selfie.c -o selfie.m
+```
+
+The command produces a file called `selfie.m` that contains machine code compiled from `selfie.c` using the *starc* compiler in `selfie` rather than the cc compiler. The process is called *self-compilation*.
```

Comment:
```suggestion
The command produces a file called `selfie.m` that contains machine code compiled from `selfie.c` using the starc compiler in `selfie` rather than the cc compiler. The process is called self-compilation.
```